dash-pipeline/SAI/src/dashsai.cpp

```diff
@@ -789,3 +789,118 @@ sai_status_t DashSai::get(
 
     return SAI_STATUS_NOT_IMPLEMENTED;
 }
+
+/**
+ * @brief Populate default attributes.
+ *
+ * Since BMv2 dont's support default attributes, we will add to existing
+ * attributes all the ones that have default values, except those that
+ * have valid only condition not met, since then BMv2 will fail.
+ *
+ * @objectType object type for list of attributes
+ * @attr_count attributes count
+ * @attr_list attributes list
+ *
+ * @return List of attributes with possible added attributes with default values.
+ */
+std::vector<sai_attribute_t> DashSai::populateDefaultAttributes(
+        _In_ sai_object_type_t objectType,
+        _In_ uint32_t attr_count,
+        _In_ const sai_attribute_t *attr_list)
+{
+    DASH_LOG_ENTER();
+
+    // populate existing attributes
+
+    std::vector<sai_attribute_t> attrs(attr_list, attr_list + attr_count);
+
+    auto* info = sai_metadata_get_object_type_info(objectType);
+
+    if (info == nullptr)
+    {
+        DASH_LOG_ERROR("failed to get metadata info for object type %d", objectType);
+        return attrs;
+    }
+
+    // iterate over all possible attributes
+
+    for (size_t idx = 0; idx < info->attrmetadatalength; idx++)
+    {
+        auto* md = info->attrmetadata[idx];
+
+        auto* attr = sai_metadata_get_attr_by_id(md->attrid, (uint32_t)attrs.size(), attrs.data());
+
+        if (attr)
+            continue;   // attribute already exists on current attribute list
+
+        if (md->isconditional)
+            continue;   // conditional attributes should be populated by user
+
+        if (md->isreadonly)
+            continue;   // can't be read only
+
+        if (md->ismandatoryoncreate)
+            continue;   // can't be mandatory on create
+
+        if (md->defaultvaluetype == SAI_DEFAULT_VALUE_TYPE_NONE)
+            continue;   // there is no default value
+
+        if (md->isvalidonly)
+        {
+            bool haveActionAttribute = false;
+
+            for (size_t i = 0; i < md->validonlylength; i++)
+            {
+                auto *condmd = sai_metadata_get_attr_metadata(objectType, md->validonly[i]->attrid);
+
+                if (condmd && strstr(condmd->attridname, "_ATTR_ACTION"))
+                {
+                    haveActionAttribute = true;
+                    break;
+                }
+            }
+
+            if (haveActionAttribute == false)
+            {
+                // always set default attribute in this case, even if condition is not set
+                // (see github discussion https://github.com/sonic-net/DASH/pull/547)
+            }
+            else if (sai_metadata_is_validonly_met(md, (uint32_t)attrs.size(), attrs.data()) == false)
+            {
+                // attribute is valid only, but condition is not met based on current attributes
```

Comment:
i believe this check might fail if we have multiple action attributes mixed with other validonly checks.

E.g.: the ATTR_OTHER is used in one of the action, say "A0", so we will have the condition `ATTR_ACTION == A0 and ATTR_OTHER == Bar`. But for action "A1", the "ATTR_OTHER" should not be added, but will be passing the check and added here.

But I guess so far this is ok, as we don't have such cases yet. And for temporary walk around, we can add this type of attribute to all possible actions to walk around this issue.